src/backend/pipeline/cont_combiner.c

```diff
@@ -812,9 +786,18 @@ read_batch(ContQueryCombinerState *state, TupleBufferBatchReader *reader)
 
 		ExecStoreTuple(heap_copytuple(tbs->tuple->heaptup), state->slot, InvalidBuffer, false);
 		tuplestore_puttupleslot(state->batch, state->slot);
+		state->batch_hashes[count] = tbs->tuple->group_hash;
 
 		IncrementCQRead(1, tbs->size);
 		count++;
+
+		/*
+		 * This is fairly unlikely, but if it does happen we should
+		 * finish processing this batch immediately because we're storing 8-byte hashes
+		 * in memory alongside each of these tuples.
+		 */
+		if (!tuplestore_in_memory(state->batch))
```

Comment:
Not sure if I follow what's going on here? We're already allocating `batch_size * 8` bytes for each state. Whether the batch of tuples fits in memory or not shouldn't make a difference?